### PR TITLE
removing race condition in dtcontext

### LINKED ISSUE #640: race condition @ DTContext.Lock / DTContext.Unlock
**What happened**:
We have a potential race @ locking / unlocking the device mutex. (https://github.com/kubeedge/kubeedge/blob/0ddc4b3f7752fb527c73bec4d5d2adbe5f3635f2/edge/pkg/devicetwin/dtcontext/dtcontext.go#L107-L128)

EXAMPLE CASE:
- 2 messages from the same device
- start handling first message
  - run dtc.Lock('xyz')
      - success on dtc.Mutex.Lock()
      - success on deviceMutex.Lock()
      - success on dtc.Mutex.Unlock()
- start handling second message
   - run dtc.Lock('xyz')
      - success on dtc.Mutex.Lock()
      - bounded waiting on deviceMutex.Lock() # beause the first message has this mutex
- continue handling first message
  - run dtc.Unlock('xyz')
     - bounded waiting on dtc.Mutex.Unlock() # because the second message has the mutex

**What you expected to happen**:
no race condition

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
not relevant

### PR DESCRIPTION
**Pull request by subpathdev**
**What type of PR is this?**
 /kind bug

**What this PR does / why we need it**:
fix race condition which is described in #640 

**Which issue(s) this PR fixes:
Fixes #640 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by subpathdev** [comment]
cc @sids-b

File: edge/pkg/devicetwin/dtcontext/dtcontext.go [hunk 1]
```diff
@@ -107,9 +107,8 @@ func (dtc *DTContext) GetMutex(deviceID string) (*sync.Mutex, bool) {
 func (dtc *DTContext) Lock(deviceID string) bool {
 	deviceMutex, ok := dtc.GetMutex(deviceID)
 	if ok {
-		dtc.Mutex.Lock()
+		dtc.Mutex.RLock()
 		deviceMutex.Lock()
-		dtc.Mutex.Unlock()
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtcontext/dtcontext.go]: @subpathdev can we have a RUnlock here ?

File: edge/pkg/devicetwin/dtcontext/dtcontext.go [hunk 2]
```diff
@@ -119,9 +118,8 @@ func (dtc *DTContext) Lock(deviceID string) bool {
 func (dtc *DTContext) Unlock(deviceID string) bool {
 	deviceMutex, ok := dtc.GetMutex(deviceID)
 	if ok {
-		dtc.Mutex.Lock()
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtcontext/dtcontext.go]: can we have a RLock here ?

**Comment by subpathdev** [comment]
Why we should directly unlock the dtc.Mutex? (I think that is the main point here) 

What are the benefits from leaving away:
  - dealMembershipDetail can only run if no reader is running (=> when dealMembershipDetail is running no changes will be made on any device)

What are the losses about this?
  - dealMembership must wait till all readers have unlocked the RWMutex

**Comment by sids-b** [comment]
> Why we should directly unlock the dtc.Mutex? (I think that is the main point here)
> 
> What are the benefits from leaving away:
> 
> * dealMembershipDetail can only run if no reader is running (=> when dealMembershipDetail is running no changes will be made on any device)
> 
> What are the losses about this?
> 
> * dealMembership must wait till all readers have unlocked the RWMutex

Sounds Fair

**Comment by sids-b** [comment]
cc @rohitsardesai83 for review

**Comment by sids-b** [comment]
cc @m1093782566 for review

**Comment by rohitsardesai83** [comment]
@subpathdev , please mark conversations as resolved if you have addressed them.

**Comment by subpathdev** [comment]
@rohitsardesai83 sry I will do this in the future

**Comment by sids-b** [comment]
@subpathdev , please give me sometime. I am thinking that if readers keep on acquiring lock on after the other, dealMembershipDetail may acquire the lock after a long time. Give me sometime to think if something can be done to solve this.

**Comment by subpathdev** [comment]
@sids-b please share your thoughts. It is a pleasure to resolve the problem with you.

If I understand the problem correct you are thinking that the read locks stop executing dealMembershipDetail for a very long time, because they are acquire the read lock again and again and ...?

This is problem should be fixed by sync.RWLock. Here an output of the documentation (https://golang.org/pkg/sync/#RWMutex):
> If a goroutine holds a RWMutex for reading and another goroutine might call Lock, no goroutine should expect to be able to acquire a read lock until the initial read lock is released. In particular, this prohibits recursive read locking. This is to ensure that the lock eventually becomes available; a blocked Lock call excludes new readers from acquiring the lock.

If I am understanding the documentation correctly, after acquire a rwmutex.Lock() no new read access on the critical section will be granted till rwmutex.Unlock() is called.
To make it concrete if dealMembershipDetail runs "dtcontex".LockAll() all new readers have to wait till dealMembershipDeatil runs "dtcontext".UnlockAll().

On the other hand we process only one message on one rwmutex.RLock call and unlock the read after finishing the process.

**Comment by sids-b** [comment]
Thanks @subpathdev

**Comment by kevin-wangzefeng** [comment]
Thanks for the fix!
/lgtm
/approve

**Pull request merged**